### PR TITLE
Catching exceptions from module

### PR DESCRIPTION
**Pull request by tomas862**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.6.x
| Description?  | in module exception is thrown and in core exception is catched and handled. See :arrow_double_down:  for sample usage within module
| Type?         | improvement
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/13490
| How to test?  | no testing data provided yet.

in module:

```php
throw \PrestaShop\PrestaShop\Core\Exception\ModuleException::buildFromArray(
            [
                $this->getTranslator()->trans('First error from module', [], 'Modules.Democqrshooksusage.Admin'),
                $this->getTranslator()->trans('Second error from module', [], 'Modules.Democqrshooksusage.Admin'),
            ]
        );
```

in core controller:

```php
catch (ModuleException $e) {
            $this->setModuleErrorMessages($e);
        }
```

Result:

Screenshot (2)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/14161)

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php
Comment (matks) [inline src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php]:
```suggestion
        if (!empty($moduleErrors)) {
```

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [inline src/Core/Exception/ModuleException.php:34]: `ModuleException` seems like a very generic exception, how about `ModuleHookException`? @matks wdyt?

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [inline src/Core/Exception/ModuleException.php]: I'd name it `translatedErrors` to make it more clear.

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [inline src/Core/Exception/ModuleException.php]: Are you sure setter is needed here?

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [inline src/Core/Exception/ModuleException.php]: Maybe `buildWithMessages` or just `withMessages`? Wdyt?

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [inline src/Core/Exception/ModuleException.php]: Although I'm not sure this static factory is useful, is it?

Comment (tomas862) [reply src/Core/Exception/ModuleException.php]: For me it also seems quite a strange way to do it. @matks, wdyt? You have suggested this approach I think the reason was to make module developers a simplified version of how to raise an exception?

File: src/Core/Exception/ModuleException.php
Comment (matks) [reply src/Core/Exception/ModuleException.php:34]: But we could use it outside of a hook 🤔if we use SF services override for example

File: src/Core/Exception/ModuleException.php
Comment (matks) [reply src/Core/Exception/ModuleException.php]: Because I wanted to allow module developers to build a `ModuleException` with either 1 or multiple error messages for maximum flexibility (with a constructor `public function __construct($error, ...)`

I expect most module developers to throw one exception for each error which means
```
throw ModuleException($error);
```
so `buildFromArray()` was a helper function to be able to deal with an array of errors

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [reply src/Core/Exception/ModuleException.php:34]: Well, `ModuleException` could be the base exception for other module related features.

File: src/Core/Exception/ModuleException.php
Comment (matks) [reply src/Core/Exception/ModuleException.php]: Also we might consider adding a `addError()` function for developers which want to build the ModuleException gradually and throw it at the end 🤔 ? or is it too much ?

**Comment by Quetzacoalt91** [comment]
Hi @tomas862,

We're close to release PrestaShop 1.7.6.0, only regressions from 1.7.5.x will be merged on this branch. I advise you to switch on the `develop` branch. :)

**Comment by matks** [comment]
> Hi @tomas862,
> 
> We're close to release PrestaShop 1.7.6.0, only regressions from 1.7.5.x will be merged on this branch. I advise you to switch on the `develop` branch. :)

It's a PR to allow modules on new migrated pages to handle errors properly 😅is it a big issue ?

**Comment by matks** [comment]
Also the PR aims to fix https://github.com/PrestaShop/PrestaShop/issues/13490 which is on 1.7.6 roadmap 😉

**Comment by kpodemski** [comment]
Oh come on... this needs to be merged into 1.7.6...

File: src/Core/Exception/ModuleException.php
Comment (sarjon) [inline src/Core/Exception/ModuleException.php]: `Throwable` is part of PHP 7, thus should not be used in 1.7.6 PS.

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/SupplierController.php
Comment (rokaszygmantas) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/SupplierController.php:40]: namespace was unused

**Comment by rokaszygmantas** [comment]
Updated this PR - cleaned up `ModuleException` class, and now migrated controllers are catching it in every relevant create/update action. Also introduced `ModuleHookException`, which extends `ModuleException`, which will allow us to control exception scope in the future.

Created a small module to test it - [link to module](https://github.com/rokaszygmantas/hookerrortest)
1. Install this module
2. Go to edit a customer, save the form. You should see multiple error messages: https://prnt.sc/o1crqq
3. Go to add new customer, fill mandatory fields, save the form. You should see one error message that's coming from the module: https://prnt.sc/o1ct8z

ping @matks

**Review by matks — DISMISSED** [review]

File: src/Core/Exception/ModuleException.php
Comment (PierreRambaud) [inline src/Core/Exception/ModuleException.php]:
```suggestion
    public function __construct($message = '', $code = 0, $previous = null)
```

File: src/Core/Exception/ModuleException.php
Comment (PierreRambaud) [inline src/Core/Exception/ModuleException.php:62]: Never used

Comment (rokaszygmantas) [reply src/Core/Exception/ModuleException.php:62]: it is meant to be used in modules :)

**Comment by matks** [comment]
> Oh come on... this needs to be merged into 1.7.6...

Hi @kpodemski, I can understand your frustration but rest assured that we will not ship 1.7.6 with a proper error system for modules !

However after discussing this, @eternoendless and I agree that this PR is a new feature and consequently cannot be merged into 1.7.6.x as we are in feature freeze and really, really close to release date. The risk is way too high.

That is why we are trying to find an eligible workaround. https://github.com/PrestaShop/PrestaShop/pull/14206 should do the job (now waiting for @eternoendless review): it focuses on fixing the issue (modules error unable to be displayed correctly) and is small enough not to introduce a risk for the incoming 1.7.6.0 release.

Once the PR workaround has been merged, we'll sit back and see what we do about this PR. Because the workaround might replace it (fully or partially) or require us to update it. When this is done, we'll merge this PR in `develop` for next version 1.7.7

**Comment by matks** [comment]
This matter was finally handled in https://github.com/PrestaShop/PrestaShop/pull/14239

**Pull request closed without merging**
